### PR TITLE
add parser options for dynamic import

### LINKED ISSUE #5029: Consider adding dynamic import's webpackMode as a configuration option, or eval the comment contents to get value
**Do you want to request a *feature* or report a *bug*?**

Feature Request
Add webpackMode as a config. option OR eval comment in dynamic import before using it's value.

**What is the current behavior?**
webpackMode defaults to "lazy", it can be overridden through a comment as /* webpackMode: "eager" */

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**
Ability to add webpackMode in config file, so preferred value in order of preference would be comment.webpackMode, config.webpackMode, webpackDefault.webpackMode

**If this is a feature request, what is motivation or use case for changing the behavior?**

Hot Module Replace works well with asynchronously imports. However, Hot Reload is unable to reload state for such code. Example: Using async. loaded components with React Hot Loader, where the component's are not refreshed. Only suitable approach seems to require loading code synchronously in development. 

Adding webpackMode in config. could be an approach to fix this. i.e. Not specifying webpackMode in comment to use "lazy" in prod. and "eager" in development.

Alternative approach would be to eval webpackMode value. i.e. 
/* webpackMode: process.env.NODE_ENV == "production" ? "lazy" : "eager" */

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

Node 6.9.1
NPM 4.1.1
Webpack 2.6.1

### LINKED ISSUE #11012: Allow global config for webpackMode
## Feature request

**What is the expected behavior?**

I expect to globally change the webpackMode of `import()` statements

**What is motivation or use case for adding/changing the behavior?**

I am in a situation in which I need `webpackMode: "lazy"` for client rendering and `webpackMode : "eager"` for server rendering. To have this dual behaviour currently I need to test some global env variable (e.g: `process.env.IS_SERVER`) for every `import()` statement.

```javascript
// written on the spot, don't expect to actually work
function loadAndUseACertainModule() {
  let mod;
  if (process.env.IS_SERVER === 'true') {
    mod = import(/*webpackMode: "eager"*/ './mod') 
  } else {
    mod = import(/*webpackMode: "lazy"*/ './mod') 
  }
  mod.then(...)
}
``` 

Hope you can see that doing this for every lazily loaded module is not ideal

**How should this be implemented in your opinion?**

A config in `webpack.config.js` could sit in between the default and the magic comment.
No idea about the config name and position.
This new config value will be used instead of the default if the magic comment is not present.

**Are you willing to work on this yourself?**
yes

---
Edit: apologies in advance if there's already a way of changing the webpackMode of import() statements without the magic comment already; Searching "eager" on the docs only revealed the magic comment bit

### PR DESCRIPTION
**Pull request by vankop**
**What kind of change does this PR introduce?**
feature
- add dynamicImportMode for ContextModule
- add dynamicImportPrefetch for prefetch order
- add dynamicImportPreload for preload order
closes #5029  
closes #11012 

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
- add `parser.javascript.dynamicImportMode` enum of `["eager", "weak", "lazy", "lazy-once"]`
- add `parser.javascript.dynamicImportPrefetch` boolean flag for prefetch order
- add `parser.javascript.dynamicImportPreload` boolean flag for preload order

this options will be used if there is no magic comments in import() (`import(/* webpackMode: "sync" */"blabla")`).

### CODE REVIEW

**Review by alexander-akait — APPROVED** [review]
Awesome :star:

File: declarations/WebpackOptions.d.ts [hunk 1]
```diff
@@ -2929,6 +2929,24 @@ export interface JavascriptParserOptions {
...
+	/**
+	 * Specifies global mode for dynamic import.
+	 */
+	dynamicImportMode?:
+		| "sync"
+		| "eager"
+		| "weak"
+		| "async-weak"
+		| "lazy"
+		| "lazy-once";
```
Comment (sokra) [inline declarations/WebpackOptions.d.ts:2941]: sync is not possible for `import()`

and `async-weak` is internal. `weak` will lead to `async-weak` automatically when using `import()`

`lazy-once` is only available for an `import()` with expression.

So this should be `eager | weak | lazy | lazy-once`

And `lazy-once` should be accepted for import() without expression too, and behave as `lazy`

File: declarations/WebpackOptions.d.ts [hunk 2]
```diff
@@ -2929,6 +2929,24 @@ export interface JavascriptParserOptions {
...
+	/**
+	 * Specifies global prefetch for dynamic import.
+	 */
+	dynamicImportPrefetch?: boolean;
```
Comment (sokra) [inline declarations/WebpackOptions.d.ts:2945]: Maybe these should be `boolean | number` to allow to specify a default priority.

**Review by sokra — CHANGES_REQUESTED** [review]

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

File: declarations/WebpackOptions.d.ts (hunk 1 above)
Comment (TongDaDa) [reply declarations/WebpackOptions.d.ts]: Awesome

**Pull request merged**